JSON parsing fix from steem PR 2178

pmconrad opened the pull request.

pmconrad committed:
JSON parsing fix from steem PR 2178

abitmore commented:
And please add this one: https://github.com/EOSIO/eos/commit/9aa838e9a7ebdc69766ed1f1e465465d52f921e9

abitmore commented:
In order to apply the EOS json patch, need to remove `variant json::from_stream( buffered_istream& in, parse_type ptype )`, otherwise it doesn't compile.

pmconrad committed:
Minor optimization

pmconrad commented:
Picked https://github.com/EOSIO/eos/commit/842e12fb3cbf7c939762458cb125a7b203e64110

Wrt https://github.com/EOSIO/eos/commit/9aa838e9a7ebdc69766ed1f1e465465d52f921e9 :

This patch cannot be applied, because variant_from_stream is called with different stream types which have inconsistent EOF handling. For example, fc::istream doesn't have an eof(), instead, *most* implementations/subclasses of fc::istream throw an eof_exception on EOF. fc::stringstream has an eof() method *and* throws. All streams derived from boost::asio throw and don't have an eof(). fc::ifstream has eof() and a read() method that throws and a read_some() method that doesn't, thus violating the contract of fc::istream. fc::json::from_file calls variant_from_stream with a boost::filesystem::ifstream that only has an eof().

This is a horrible mess. Investigating.

pmconrad committed:
Added unit tests for stringstream, [io]fstream and buffered_[io]stream

pmconrad committed:
Added first json test case

pmconrad committed:
Fixed from_file

pmconrad committed:
Deduplicate some code

abitmore commented:
In EOS, before applying the patch, [some code](https://github.com/EOSIO/eos/blob/9aa838e9a7ebdc69766ed1f1e465465d52f921e9/libraries/fc/src/io/json.cpp#L815-L832) got commented out, apparently the code is not being used. Then it compiles.

@pmconrad do you prefer we merge the patch now, or merge after you finished investigation?

pmconrad commented:
Give me some more time please. Unless you think it's a critical issue.

pmconrad committed:
Added json write/validate/read test

pmconrad committed:
Code cleanups and simplifications

pmconrad committed:
Code deduplication

pmconrad commented:
Fixed json::from_file, added unit tests. Please review.

It seems the graphene codebase only uses one of the configurable parsers, so there is some superfluous stuff in there.

pmconrad commented:
Fix is incomplete (json::variants_from_string uses relaxed parser)

pmconrad committed:
Avoid legacy_generator

pmconrad committed:
Make unused parsers compile-time optional

pmconrad committed:
Include variants_from_string in parser test

pmconrad committed:
Fixed relaxed parser wrt "" input

abitmore commented on src/io/json.cpp:
```diff
@@ -797,23 +786,9 @@ namespace fc
-          default:
-              FC_ASSERT( false, "Unknown JSON parser type {ptype}", ("ptype", ptype) );
-      }
+      fc::istream_ptr in( new fc::ifstream( p ) );
```
Need `std::ios::binary` here.

abitmore replied:
By the way, need `std::ios::binary` in `from_file(...)` below as well.

pmconrad replied:
ifstream always calls open with std::io::binary, see https://github.com/bitshares/bitshares-fc/blob/c16bb206a39b09b03423088c831bc193de89056c/src/io/fstream.cpp#L64

abitmore replied:
Got it. Thanks. (Just noticed the 2 comments above are talking about same code).

pmconrad committed:
Applied variant_from_stream fix from regular to relaxed

abitmore reviewed (dismissed):
Mostly fine. Need to fix another issue as discussed in Telegram.

abitmore commented:
By the way, in `void to_stream( T& os, const variant& v, json::output_formatting format )`, when handling `variant::int64_type`, didn't check for negative value.

pmconrad committed:
Added test case for recursion depth 240

pmconrad committed:
Fix for recursion depth limitation

pmconrad committed:
Added broken_nul_parser to preserve previous behaviour

abitmore commented on include/fc/io/json_relaxed.hpp:
```diff
@@ -686,60 +625,52 @@ namespace fc { namespace json_relaxed
-              FC_THROW_EXCEPTION( parse_error_exception, "Unexpected char '${c}' in \"${s}\"",
-                                 ("c", c)("s", stringFromToken(in)) );
-         }
+         case '"':
```
How about single quotes? According to `stringFromStream(...)`, If `strict` is `false`, need to support it and raw strings etc.

abitmore commented on include/fc/io/json_relaxed.hpp:
```diff
@@ -686,60 +625,52 @@ namespace fc { namespace json_relaxed
+         case 0:
+         default:
+            FC_THROW_EXCEPTION( parse_error_exception, "Unexpected char '${c}' in \"${s}\"",
+                                ("c", c)("s", stringFromToken(in)) );
```
How about `'\'`? According to other code, unquoted strings are allowed in non-strict mode.

abitmore commented on src/io/json.cpp:
```diff
@@ -188,18 +190,17 @@ namespace fc
                continue;
             }
```
So ',' can be omitted, e.g. replaced by spaces?
And allow/ignore something like `,     , ,,,`?

pmconrad replied:
Noticed that too. Question is, how much of the original behaviour do we want to change? (Same for other comments above.)

abitmore commented on src/io/json.cpp:
```diff
@@ -276,6 +290,7 @@ namespace fc
                  if (dot)
                     FC_THROW_EXCEPTION(parse_error_exception, "Can't parse a number with two decimal places");
                  dot = true;
+                 [[fallthrough]];
```
This attribute is since C++17. I heard Visual Studio doesn't support it but will generate an error (I didn't verify).

pmconrad replied:
Oh, I see. Unknown attributes are to be ignored - but only since C++17. :-/

abitmore commented on src/io/json.cpp:
```diff
@@ -299,16 +314,20 @@ namespace fc
       }
       catch (fc::eof_exception&)
-      {
+      { // EOF ends the loop
```
'\0' will ends the loop as well, but hasn't been handled.
E.G. if the string is `"-\0[whatever]"`, `ss.str()` would be `"-"`, so it wouldn't get caught by the "obviously wrong things" check below, finally an exception will be thrown on `return to_int64(str);`.

pmconrad replied:
Yes, but the same is true for "-[whatever]".
The logic here is: a number is a sequence of digits 0-9, possibly separated once by a dot, and possibly prefixed with a single dash. It doesn't matter if the [whatever] sequence starts with a \0 or something else.
Will add "-" to the list of obviously wrong things.

abitmore commented on src/io/json.cpp:
```diff
@@ -616,27 +590,19 @@ namespace fc
-              int64_t i = v.as_int64();
               if( format == json::stringify_large_ints_and_doubles &&
-                  i > 0xffffffff )
+                  v.as_int64() > 0xffffffff )
```
Do we need to quote it if `v.as_int64() < -(int64_t)0xffffffff`?

pmconrad replied:
I think we should.

abitmore commented on src/io/json.cpp:
```diff
@@ -616,27 +590,19 @@ namespace fc
               return;
          case variant::int64_type:
-         {
-              int64_t i = v.as_int64();
```
Why change this?

pmconrad replied:
Style, mostly.
IMO these blocks are clumsy, and they're only required so you can declare the "i" variable within the case.

pmconrad committed:
Minor fixes

abitmore reviewed (approved):
Looks good.

abitmore commented on src/io/json.cpp:
```diff
@@ -627,27 +596,19 @@ namespace fc
-              int64_t i = v.as_int64();
               if( format == json::stringify_large_ints_and_doubles &&
-                  i > 0xffffffff )
+                  ( v.as_int64() > 0xffffffff || v.as_int64() < -int64_t(0xffffffff) ) )
```
Actually I think the range should be `[-0x7fffffff, 0x7fffffff]`

pmconrad replied:
AFAIK the reason for this is the number representation in some scripting languages like JavaScript, which is based on IEEE doubles with 52 bits of precision. 2^32-1 should be safe, as well as -(2^32-1).

abitmore replied:
... unless doing bit-shifting.
https://stackoverflow.com/questions/307179/what-is-javascripts-highest-integer-value-that-a-number-can-go-to-without-losin

pmconrad replied:
The string representation will not enable bit shifting on large integers either.

abitmore replied:
> The string representation will not enable bit shifting on large integers either.

True, but the point is, if it represents as a number, it should enable bit shifting, so 2^31 -1 is the max safe number.

pmconrad committed:
Make broken_nul_parser usable

abitmore commented:
@pmconrad `to_stream()` has recursion issue.

pmconrad committed:
Added max_depth parameter to all to_/from_ methods

pmconrad committed:
Stringify numbers >MAXINT or <MININT

abitmore commented on src/io/json.cpp:
```diff
@@ -588,8 +586,9 @@ namespace fc
-   void to_stream( T& os, const variant& v, json::output_formatting format )
+   void to_stream( T& os, const variant& v, json::output_formatting format, uint32_t max_depth )
    {
+      FC_ASSERT( max_depth > 0, "Too many nested objects!" );
```
There is an issue related to this, although not necessarily caused by it. For example, among this code:

```
object o( a_nested_object );
variant a;
try {
   to_stream( nested_object, a );
} FC_CAPTURE_AND_RETHROW ( (o) );
```

`FC_CAPTURE_AND_RETHROW ();` will try to jsonify `o` which is the nested object, so it will throw another exception.

We need to carefully check all `catch`es and make sure they don't throw again.

pmconrad committed:
Added test case for exception in exception handling

pmconrad committed:
Moved format_string from variant.cpp to string.cpp (it is declared in string.hpp), added handling of recursion errors

abitmore commented:
@pmconrad something is broken. Unable to make witness_node. Maybe related to the code handling '$' in `format_string`?
```
[ 96%] Building CXX object libraries/egenesis/CMakeFiles/graphene_egenesis_full.dir/egenesis_full.cpp.o
/app/bts-dev/build/libraries/egenesis/egenesis_full.cpp:43:38: error: unable to find numeric literal operator ‘operator""$’
 static const char genesis_json_array[941747$][40$+1] =
```

pmconrad commented:
Sorry, didn't test that.
IMO the fault lies with libraries/egenesis/egenesis_*.cpp.tmpl - in some places they use placeholders of the form ${x}$ instead of the correct ${x}. The previous version of format_string would skip a $ char if it wasn't followed by a {, which I think is wrong.
